app.py

```diff
@@ -1,37 +1,118 @@
+    
+    os.environ['OAUTHLIB_INSECURE_TRANSPORT'] = '1'
+
+    app.secret_key = os.urandom(24)
+
+    login_disabled = os.getenv('LOGIN_DISABLED') == 'TRUE'
+    app.config['LOGIN_DISABLED'] = login_disabled
+    
+    login_manager = LoginManager()
+
+    client_id = os.getenv('CLIENT_ID')
+    client_secret = os.getenv('CLIENT_SECRET')
+    client = WebApplicationClient(client_id)
+
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        redirect_uri = client.prepare_request_uri(
+            "https://github.com/login/oauth/authorize"
+        )
+        return redirect(redirect_uri)
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id)
+
+    login_manager.init_app(app)
+
+    @app.route("/login/callback", methods=["GET"])
+    def callback():
+        request_url, token_headers, token_body = client.prepare_token_request(
+            token_url="https://github.com/login/oauth/access_token",
+            authorization_response=request.url,
+            client_secret=client_secret
         )
+        token_response = requests.post(
+            request_url,
+            headers=token_headers,
+            data=token_body
+        )
+        client.parse_request_body_response(
+            token_response.content.decode()
+        )
+
+        user_info_request_url, user_info_request_headers, user_info_request_body = client.add_token(
+            uri='https://api.github.com/user',
+        )
+        user_info_response = requests.get(
+            user_info_request_url,
+            data=user_info_request_body,
+            headers=user_info_request_headers
+        )
+
+        github_username = user_info_response.json()['login']
+        user = User(github_username)
+        login_success = login_user(user)
+
+        if login_success:
+            return redirect(url_for('index'))
+        else:
+            return "Unauthorised", 403   
 
     @app.route('/')
-    def index():   
-        items = []
-        trello_cards = client.get_all_cards_for_board(os.getenv('BOARD_ID'))
-        for card in trello_cards:
-            items.append(TrelloItem.from_raw_trello_card(card))
-        item_view_model = ViewModel(items)
-        return render_template('index.html', view_model=item_view_model)
-
-    @app.route('/', methods=['POST'])
-    def add_todo_card():
-        new_card = request.form.get('todoitem')
-        add_card = client.add_card_to_list(os.getenv('TODO_LIST_ID'),new_card)
+    @login_required
+    def index():
+        user: User = current_user
+        readonly = (not login_disabled) and user.get_role() == UserRole.Reader
+        readonly = False
+        items = get_items(collection)
+
+        return render_template('index.html', view_model=ViewModel(items,readonly))
+
+    @app.route('/items/new', methods=['POST'])
+    @login_required
+    def add_item():
+        title = request.form['title']
+        add_new_item(collection, title)
```

Comment:
Here you've not got any checks that determine if the user has permission to hit this endpoint; so while we hide the functionality we don't actually stop them achieving the goal if they can guess/work out/observe your endpoints. We need to also add server side verification that the user has the appropriate permissions

This could be added per endpoint:
```
if current_user.get_role() != UserRole.Writer:
    return "Unauthorised", 403
```

Or we could wrap it in a decorator ([e.g. see the do_twice decorator here for inspiration](https://realpython.com/primer-on-python-decorators/#who-are-you-really)) to make that a bit more reusable, and just add the annotation to the relevant functions


Note that for adjusting your end-to-end test, you may want to either allow users through if login is disabled, or configure the anonymous user